aten/src/ATen/native/mps/operations/Quantized.mm

```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+        }
+
+        a_ptr += BLOCK_SIZE_K;
+        b_ptr += BLOCK_SIZE_K;
+
+        threadgroup_barrier(mem_flags::mem_threadgroup);
+
+        // load matrices from threadgroup memory and conduct outer products
+        threadgroup T      * lsma = (sa + THREAD_MAT_M * SG_MAT_SIZE * (sgitg / 2));
+        threadgroup bfloat * lsmb = (sb + THREAD_MAT_N * SG_MAT_SIZE * (sgitg % 2));
+
+/**
+Outer product:
+              K
+       ────────────►
+     8    for loop              8   8
+   ┌───┬───┬───┬───┐          ┌───┬───┬───┬───┬───┬───┬───┬───┐
+ 8 │+++│   │   │   │      │  8│+++│+++│+++│+++│###│###│###│###│
+   ├───┼───┼───┼───┤      │   ├───┼───┼───┼───┼───┼───┼───┼───┤
+   │+++│   │   │   │      │   │   │   │   │   │   │   │   │   │
+   ├───┼───┼───┼───┤      │ K ├───┼───┼───┼───┼───┼───┼───┼───┤
+   │###│   │   │   │      │   │   │   │   │   │   │   │   │   │
+   ├───┼───┼───┼───┤      │   ├───┼───┼───┼───┼───┼───┼───┼───┤
+   │###│   │   │   │      │   │   │   │   │   │   │   │   │   │
+   └───┴───┴───┴───┘      ▼   └───┴───┴───┴───┴───┴───┴───┴───┘
+                       for loop
+    + simdgroup 0,1                + simdgroup 0,2
+    # simdgroup 2,3                # simdgroup 1,3
+ */
+        #pragma unroll(4)
+        for (short ik = 0; ik < BLOCK_SIZE_K / 8; ik++) {
+            #pragma unroll(4)
+            for (short i = 0; i < 4; i++) {
+                simdgroup_load(mb[i], lsmb + SG_MAT_SIZE * i);
+            }
+            simdgroup_barrier(mem_flags::mem_none);
+            #pragma unroll(2)
+            for (short i = 0; i < 2; i++) {
+                simdgroup_load(ma[i], lsma + SG_MAT_SIZE * i);
+            }
+
+            lsma += BLOCK_SIZE_M / SG_MAT_ROW * SG_MAT_SIZE;
+            lsmb += BLOCK_SIZE_N / SG_MAT_ROW * SG_MAT_SIZE;
+
+            #pragma unroll(8)
+            for (short i = 0; i < 8; i++){
+                simdgroup_multiply_accumulate(c_res[i], ma[i/4], mb[i%4], c_res[i]);
+            }
+        }
+    }
+
+    /**
+ * Each sgitg 0,1,2,3 handles 2x4 8x8.
+    8   8
+  ┌───┬───┬───┬───┬───┬───┬───┬───┐
+ 8│ 0 │ 0 │ 0 │ 0 │ 1 │ 1 │ 1 │ 1 │
+  ├───┼───┼───┼───┼───┼───┼───┼───┤
+  │ 0 │ 0 │ 0 │ 0 │ 1 │ 1 │ 1 │ 1 │
+  ├───┼───┼───┼───┼───┼───┼───┼───┤
+  │ 2 │ 2 │ 2 │ 2 │ 3 │ 3 │ 3 │ 3 │
+  ├───┼───┼───┼───┼───┼───┼───┼───┤
+  │ 2 │ 2 │ 2 │ 2 │ 3 │ 3 │ 3 │ 3 │
+  └───┴───┴───┴───┴───┴───┴───┴───┘
+
+   scale: 8 x BLOCK_SIZE_N, starting from sa. Each sgitg handles 4 8x8 diagonal matrix.
+    8   8
+  ┌───┬───┬───┬───┬───┬───┬───┬───┐
+ 8│   │   │   │   │   │   │   │   │
+  └───┴───┴───┴───┴───┴───┴───┴───┘
+ */
+
+    threadgroup float * temp_str = ((threadgroup float *)shared_memory) \
+                                  + 32 * (sgitg&1) + (16 * (sgitg>>1)) * BLOCK_SIZE_N;
+    for (int i = 0; i < 8; i++) {
+        int block_start = 4 * 8 * (sgitg & 1) + (i % 4) * 8;
+        threadgroup float * temp_scale = (threadgroup float *)sb + block_start;
+        threadgroup float * scale_itr = temp_scale;
```

Comment:
what is an "itr"? "iterator"? ("iter" would be OK per longstanding programming convention)